### PR TITLE
Fix subStringEnd calculation

### LINKED ISSUE #1842: Batch delete fails with InvalidOperation error
### Which service(blob, file, queue, table) does this issue concern?

table

### Which version of the Azurite was used?

3.22.0

### Where do you get Azurite? (npm, DockerHub, NuGet, Visual Studio Code Extension)
npm
### What's the Node.js version?
v16.19.0
### What problem was encountered?
We're testing durable functions on a linux host with function core tools. When the durable function receives a Purge Instance History command, it starts a batch delete of the instance history. This batch delete fails with InvalidOperation being returned.
### Steps to reproduce the issue?

- Create a durable function and run it on a linux host with function core tools. As long as it has orchestrator and activity functions should do.
- Execute Purge Instance History.

If possible, please provide the debug log using the -d parameter, replacing \<pathtodebuglog\> with an appropriate path for your OS, or review the instructions for docker containers:

```
2023-03-17T15:07:07.172Z 5b1743a1-6387-4389-83c6-419adf0d1ac7 debug: TableHandler:batch() Raw request string is "--batch_e6bedae0-33a0-4875-bf7a-dc3963071819\nContent-Type: multipart/mixed; boundary=changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='0000000000000000') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A06%3A18.3075721Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='0000000000000001') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A06%3A18.3075732Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='0000000000000002') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A06%3A18.3075737Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='0000000000000003') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A06%3A18.3075742Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='0000000000000004') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A06%3A19.3496077Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='0000000000000005') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A06%3A19.3496080Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='0000000000000006') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A06%3A19.3496083Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='0000000000000007') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A06%3A19.3496085Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='0000000000000008') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A06%3A19.4196778Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='0000000000000009') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A06%3A19.4196781Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='000000000000000A') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A06%3A19.4196783Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='000000000000000B') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A06%3A19.4196786Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='000000000000000C') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A07%3A02.1774357Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='000000000000000D') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A07%3A02.1774360Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='000000000000000E') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A07%3A02.1774364Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='000000000000000F') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A07%3A02.1774367Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nDELETE http://127.0.0.1:10002/devstoreaccount1/TestHubNameHistory(PartitionKey='00000000EDGC5674',RowKey='sentinel') HTTP/1.1\nAccept: application/json;odata=minimalmetadata\nContent-Type: application/json\nDataServiceVersion: 3.0;\nIf-Match: W/\"datetime'2023-03-17T15%3A07%3A02.1774369Z'\"\n\n--changeset_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf--\n--batch_e6bedae0-33a0-4875-bf7a-dc3963071819--\n"
2023-03-17T15:07:07.173Z 5b1743a1-6387-4389-83c6-419adf0d1ac7 debug: TableHandler:batch() Raw response string is "--batchresponse_e6bedae0-33a0-4875-bf7a-dc3963071819\r\nContent-Type: multipart/mixed; boundary=changesetresponse_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\r\n\r\n--changesetresponse_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf\r\nContent-Type: application/http\r\nContent-Transfer-Encoding: binary\r\n\r\nHTTP/1.1 400 Bad Request\r\nContent-ID: 1\r\nDataServiceVersion: 3.0;\r\nContent-Type: application/json;odata=minimalmetadata;charset=utf-8\r\n\r\n{\"odata.error\":{\"code\":\"InvalidOperation\",\"message\":{\"lang\":\"en-US\",\"value\":\"0:\\nRequestId:5b1743a1-6387-4389-83c6-419adf0d1ac7\\nTime:2023-03-17T15:07:07.173Z\"}}}\r\n--changesetresponse_f3679e9d-3491-4ba7-95f3-59a8e24cd7bf--\r\n--batchresponse_e6bedae0-33a0-4875-bf7a-dc3963071819--\r\n"
```

Please be sure to remove any PII or sensitive information before sharing!  
The debug log will log raw request headers and bodies, so that we can replay these against Azurite using REST and create tests to validate resolution.

### Have you found a mitigation/solution?

The deserializeBatchRequest method in TableBatchSerialization has a bug where `DELETE` operations will flow to the else block after `jsonOperationBody != null`. In this block subStringEnd is set to `subRequest.length - 4`

The issue with this is on linux the line endings are `\n\n` and when four characters are subtracted it ends up removing the `'"` at the end of the Etag, so it becomes `W/"datetime'2023-03-17T14:04:43.2505080Z`.

Later on, during the `deleteEntity` portion of the operation the Etag then fails the regex in `isEtagValid`.

I've tested changing the `subStringEnd` line to `subStringEnd = subRequest.length - (HTTP_LINE_ENDING.length * 2);` and fixed the issue.

### PR DESCRIPTION
**Pull request by haodeon**
Fixes #1842

PR has test for change and updated changelog.

@microsoft-github-policy-service agree

### CODE REVIEW

**Comment by haodeon** [comment]
@microsoft-github-policy-service agree

File: src/table/batch/TableBatchSerialization.ts [hunk 1]
```diff
@@ -110,7 +110,7 @@ export class TableBatchSerialization extends BatchSerialization {
           jsonBody = jsonOperationBody[0];
         } else {
           // remove 1 \r\n
```
Comment (edwin-huber) [inline src/table/batch/TableBatchSerialization.ts:112]: The comment no longer matches what the code is doing.

File: tests/table/unit/deserialization.unit.test.ts [hunk 2]
```diff
@@ -506,4 +506,68 @@ describe("batch deserialization unit tests, these are not the API integration te
     );
     done();
   });
+
+  it("deserializes, mock table batch request containing 4 deletes correctly", (done) => {
```
Comment (edwin-huber) [inline tests/table/unit/deserialization.unit.test.ts:510]: The test in this case is only checking the ```\n``` case, could you include a mock request using the Windows line endings as well?
Thanks!

**Review by edwin-huber — CHANGES_REQUESTED** [review]
Thanks for taking the time to figure this out and submit a change. 😸 
Please look at the changes requested.

File: src/table/batch/TableBatchSerialization.ts (hunk 1 above)
Comment (haodeon) [reply src/table/batch/TableBatchSerialization.ts]: The comment was introduced in 95a932ec4403bc70faf86fd22ad75c00228a8762 which also changed `subStringEnd`.
I couldn't understand how the comment matched the code, so I avoided changing it.
The comment is now updated.

File: tests/table/unit/deserialization.unit.test.ts (hunk 2 above)
Comment (haodeon) [reply tests/table/unit/deserialization.unit.test.ts]: I have added a mock request string and test case for `\r\n`.

While implementing the new test case, I found a potential bug in https://github.com/Azure/Azurite/blob/e5b5d588435f175e5bcb956f36ec677a294119ba/src/common/batch/BatchOperation.ts#L25

My test case was failing but only for the last operation in a batch. I eventually worked out the last ETag in the batchOperationArray has an extra `\r` but only when the batch uses `\r\n\` line endings.

I checked the behaviour of deserializeBatchRequest before this PR and it is the same, the last Etag has an extra `\r`.

In the new test, when asserting the Etag header for the last operation, the expected string has an extra `\r` at the end.

Comment (haodeon) [reply tests/table/unit/deserialization.unit.test.ts]: I removed the extra `\r` from expected string in the new test. It is passing after rawHeaders has been fixed.

**Comment by haodeon** [comment]
Hey @edwin-huber or @blueww,

Is there anything else required to merge this PR?

The issue here is affecting my team daily.

**Comment by blueww** [comment]
@haodeon 
Would you please merge the latest Azurite change to the PR?
If the PR check can pass after merge the latest change, I will merge this PR.

File: ChangeLog.md [hunk 3]
```diff
@@ -31,6 +31,7 @@ Table:
 - Fixed issue for querying GUIDs using operators other than eq and ne
 - GUID queries only support persistent storage on legacy (string) format GUIDs for eq and ne operators, other operators will only evaluate newly stored entities.
 - Fixed issue with boolean values not being recognized in query if using different cased characters.
+- Fixed issue with headers length when deserializing batch deletes.
```
Comment (blueww) [inline ChangeLog.md:34]: Please move the change log under the Upcoming release.
It's currently in 3.23.0 release, which is already published without this change.

Comment (haodeon) [reply ChangeLog.md]: I have made the requested change.

**Comment by haodeon** [comment]
@blueww
Please let me know what you would like done about the one failing test.

**Comment by blueww** [comment]
@haodeon 

Thanks for the contribution!
I have rerun the test and it passed, so merge the PR.

**Pull request merged**